src/transformStateWithClones.js

```
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    const nextState = { ...currentState };
+
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(nextState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete nextState[key];
+        }
+        break;
+
+      case 'clear':
+        for (const key in nextState) {
+          delete nextState[key];
+        }
+        break;
+
+      default:
+        break;
```

Let's throw a descriptive error in default case